cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java

```diff
@@ -276,18 +341,20 @@ public boolean equals(Object o) {
 
       DatasetCacheKey that = (DatasetCacheKey) o;
 
-      return Objects.equal(this.name, that.name) &&
+      return Objects.equal(this.namespace, that.namespace) &&
+        Objects.equal(this.name, that.name) &&
         Objects.equal(this.arguments, that.arguments);
     }
 
     @Override
     public int hashCode() {
-      return Objects.hashCode(name, arguments);
+      return Objects.hashCode(namespace, name, arguments);
```

Comment:
Cause even cache key with different accessTypes can point to one dataset instance? I see it's authored by @anwar6953 . Will check with him.